linetools/spectra/xspectrum1d.py

```diff
@@ -84,92 +75,205 @@ def from_file(self, ifile, **kwargs):
           Filename
         """
         # put import here to avoid circular import with io.py
-        from .io import readspec
-        slf = readspec(ifile, **kwargs)
+        #from .io import readspec
+        from linetools.spectra import io as tio
+        reload(tio)
+        slf = tio.readspec(ifile, **kwargs)
         return slf
 
     @classmethod
     def from_spec1d(cls, spec1d):
         """ Input Spectrum1D
-        """
-        # Giddy up
-        slf = cls(flux=spec1d.flux, wcs=spec1d.wcs, unit=spec1d.unit,
-                  uncertainty=spec1d.uncertainty, mask=spec1d.mask,
-                  meta=spec1d.meta.copy())
-        slf.co = (spec1d.co if hasattr(spec1d, 'co') else None)
 
+        Avoids error array for now
+        """
+        slf = cls.from_tuple((spec1d.dispersion, spec1d.flux))
+        #slf = cls(flux=spec1d.flux, wcs=spec1d.wcs, unit=spec1d.unit,
+        #          uncertainty=spec1d.uncertainty, mask=spec1d.mask,
+        #          meta=spec1d.meta.copy())
         return slf
 
     @classmethod
     def from_tuple(cls, ituple, sort=True, **kwargs):
-        """Make an XSpectrum1D from a tuple of arrays.
+        """Make an XSpectrum1D from a tuple of numpy arrays or Quantity arrays
 
         Parameters
         ----------
         ituple : (wave,flux), (wave,flux,sig) or (wave,flux,sig,cont)
+          ndarray or Quantity array
           If wave is unitless, Angstroms are assumed
+          If flux is unitless, it is made dimensionless
         sort : bool, optional
           Sort by wavelength?
         """
-        # Units
+        # Parse wavelength
         try:
             wv_unit = ituple[0].unit
         except AttributeError:
+            iwave = ituple[0]
+            warnings.warn("Assuming wavelength unit is Angstroms")
             wv_unit = u.AA
         else:
             if wv_unit is None:
                 wv_unit = u.AA
-        uwave = u.Quantity(ituple[0], unit=wv_unit)
-        # Sort
+            iwave = ituple[0].value
+
+        # Parse flux
+        try:
+            fx_unit = ituple[1].unit
+        except AttributeError:
+            iflux = ituple[1]
+            fx_unit = u.dimensionless_unscaled
+        else:
+            if fx_unit is None:
+                fx_unit = u.dimensionless_unscaled
+            iflux = ituple[1].value
+
+        # Sort and append None
         ltuple = list(ituple)
+        for ii in range(len(ltuple),4):
+            ltuple.append(None)
         if sort:
-            srt = np.argsort(uwave)
-            uwave = uwave[srt]
-            for i in range(1, len(ituple)):
-                if ltuple[i] is not None:
-                    ltuple[i] = ituple[i][srt]
+            srt = np.argsort(iwave)
+            iwave = iwave[srt]
```

Comment:
We should probably check that `iwave` and `iflux` have the same dimensions (same for sig and cont).